docs/specification/new-spec/consensus-reactor.md

```diff
@@ -0,0 +1,239 @@
+	Proposal           Proposal
+	ProposalBlock      Block
+	ProposalBlockParts PartSet
+	LockedRound        int
+	LockedBlock        Block
+	LockedBlockParts   PartSet
+	Votes              HeightVoteSet
+	LastCommit         VoteSet 
+	LastValidators     ValidatorSet
+}    
+``` 
+
+Internally, consensus will run as a state machine with the following states:
+RoundStepNewHeight, RoundStepNewRound, RoundStepPropose, RoundStepPropose, RoundStepPrevote,       
+RoundStepPrevoteWait, RoundStepPrecommit, RoundStepPrecommitWait and RoundStepCommit.        
+
+## Peer Round State
+
+Peer round state contains the known state of a peer. It is being updated by the Receive routine of
+Consensus Reactor and by the gossip routines upon sending a message to the peer.
+
+```
+type PeerRoundState struct {
+	Height                   int64               // Height peer is at
+	Round                    int                 // Round peer is at, -1 if unknown.
+	Step                     RoundStepType       // Step peer is at
+	Proposal                 bool                // True if peer has proposal for this round
+	ProposalBlockPartsHeader PartSetHeader 
+	ProposalBlockParts       BitArray       
+	ProposalPOLRound         int                 // Proposal's POL round. -1 if none.
+	ProposalPOL              BitArray            // nil until ProposalPOLMessage received.
+	Prevotes                 cmn.BitArray        // All votes peer has for this round
+	Precommits               cmn.BitArray        // All precommits peer has for this round
+	LastCommitRound          int                 // Round of commit for last height. -1 if none.
+	LastCommit               BitArray            // All commit precommits of commit for last height.
+	CatchupCommitRound       int                 // Round that we have commit for. Not necessarily unique. -1 if none.
+	CatchupCommit            BitArray            // All commit precommits peer has for this height & CatchupCommitRound
+}
+```   
+
+## Receive method of Consensus reactor
+
+When a message is received from a peer p, it is treated differently based on the channel message belongs to.
+
+TODO: Add high level description of the logic of Receive method. 
+
+## Gossip Data Routine
+
+It is used to send the following messages to the peer: `BlockPartMessage`, `ProposalMessage` and 
+`ProposalPOLMessage` on the DataChannel. The gossip data routine is based on the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.ProposalBlockPartsHeader == prs.ProposalBlockPartsHeader and the peer does not have all the proposal parts then
+        Part = pick a random proposal block part the peer does not have 
+        Send BlockPartMessage(Height, Round, Part) to the peer on the DataChannel 
+        if send returns true, record that the peer knows the corresponding block Part
+	    Continue  
+ 
+1b) if (0 < prs.Height) and (prs.Height < rs.Height) then
```

Comment:
What if processes p1 and p4 are at height 5, and are connected over processes p2 and p3 that are at height 3 (p1 -> p2,p3 -> p4). Then p1 can't gossip messages to p4 (and vice versa) before p2 or p3 catch up to height 5. Can this be problem?

Reply:
That's right. Why would it be a problem? So long as they eventually catch up, we can start making progress again. 

It is a problem if eg validators are only connected by peers that are always lagging, but that's an issue to be dealt with elsewhere, I would think.

How else could we handle this ?